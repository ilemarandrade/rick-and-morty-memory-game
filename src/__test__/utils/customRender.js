import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import CharactersProvider from "../../providers/CharactersProvider";

const AllTheProviders = ({ children }) => {
  return (
    <BrowserRouter>
      <CharactersProvider>{children}</CharactersProvider>
    </BrowserRouter>
  );
};

const customRender = (ui, { ...options } = {}) =>
  render(ui, {
    wrapper: ({ children }) => <AllTheProviders children={children} />,
    ...options,
  });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
