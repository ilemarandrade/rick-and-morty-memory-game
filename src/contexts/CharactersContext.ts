import { createContext, useContext } from "react";
import { ICharacters } from "../models/endPointsModel";

interface IContext {
  characters: ICharacters[] | null;
}
const CharactersContext = createContext<IContext>({ characters: null });

export const useCharactersState = () => {
  const context = useContext(CharactersContext);
  if (!context) {
    throw new Error(
      "useCharactersState must be used within a CharactersProvider"
    );
  }
  return context;
};

export default CharactersContext;
