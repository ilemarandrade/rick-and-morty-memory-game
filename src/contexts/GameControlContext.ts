import { createContext, useContext } from "react";
import { ICharacters } from "../models/endPointsModel";

interface IContext {
  characters: ICharacters[] | null;
  matchesGot: number;
  setMatchesGot: (n: number) => void;
  turns: number;
  setTurns: (n: number) => void;
}

const GameControlContext = createContext<IContext>({
  characters: null,
  matchesGot: 0,
  setMatchesGot: () => {},
  turns: 0,
  setTurns: () => {},
});

export const GameControlState = () => {
  const context = useContext(GameControlContext);

  if (!context) {
    throw new Error(
      "GameControlState must be used within a GameControlProvider"
    );
  }

  return context;
};

export default GameControlContext;
