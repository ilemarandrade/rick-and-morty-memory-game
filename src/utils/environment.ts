const env = {
  API_BASE_URL: process.env.REACT_APP_API_URL,
  APP_ENV: process.env.REACT_APP_APP_ENV,
};

export default env;
