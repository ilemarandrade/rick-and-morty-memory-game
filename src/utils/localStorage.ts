export const KEY_SHOW_INTRUCTIONS = "show-intructions-key";

export const getKey = (key: string) => localStorage.getItem(key);

export const setKey = ({ key, value }: { key: string; value: string }) =>
  localStorage.setItem(key, value);
