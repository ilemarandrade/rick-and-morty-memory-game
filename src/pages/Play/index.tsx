import { useEffect, useState } from "react";
import GameBoard from "../../components/GameBoard";
import { ICharacters } from "../../components/models/endPointsModel";
import { AllNumberKeys } from "../../components/models/generics";
import { useCharactersState } from "../../contexts/CharactersContext";

const randomCharacters = (characters: ICharacters[] | null) => {
  if (characters) {
    return [...characters?.slice(0, 6), ...characters?.slice(0, 6)]?.sort(
      () => 0.5 - Math.random()
    );
  }
  return;
};

const Play = () => {
  const { characters } = useCharactersState();
  const [charactersState, setCharactersState] = useState(() => {
    const value = randomCharacters(characters)?.map(({ ...keys }, index) => ({
      ...keys,
      open: true,
      position: index,
    }));
    return value;
  });
  const closeAll = () => {
    setCharactersState(
      charactersState?.map(({ ...values }) => ({
        ...values,
        open: false,
      }))
    );
  };

  const openCard = ({ position, id }: AllNumberKeys) => {
    setCharactersState((prev) =>
      prev?.map((currentCharacter) =>
        currentCharacter.position === position
          ? { ...currentCharacter, open: true }
          : currentCharacter
      )
    );
  };

  useEffect(() => {
    setTimeout(() => {
      closeAll();
    }, 3000);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <>
      <div className="d-flex jc-space-between">
        <h2 className="mb-1-5">Aciertos:</h2>
        <h2 className="mb-1-5">Turnos:</h2>
      </div>
      {charactersState && (
        <GameBoard data={charactersState} onClickToCard={openCard} />
      )}
    </>
  );
};

export default Play;
