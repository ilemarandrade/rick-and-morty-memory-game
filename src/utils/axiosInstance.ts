import axios from "axios";
import environment from "./environment";

const instance = axios.create({
  baseURL: environment.API_BASE_URL,
  headers: {
    Accept: "application/json",
    "Content-Type": "application/json",
  },
});

export default instance;
