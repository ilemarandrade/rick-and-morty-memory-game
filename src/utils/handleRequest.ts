import { AxiosRequestConfig, Method } from "axios";
import axiosInstance from "./axiosInstance";
interface IhandleRequest<T> {
  method: Method;
  url: string;
  config?: AxiosRequestConfig;
  onSuccess: (res: { data: T }) => void;
  onError: (res: { data: T }) => void;
}

const handleRequest = <T>({
  method,
  url,
  config,
  onSuccess,
  onError,
}: IhandleRequest<T>) => {
  axiosInstance({
    method,
    url,
    ...config,
  })
    .then((res) => {
      onSuccess(res);
    })
    .catch((err) => {
      onError(err);
    });
};

export default handleRequest;
