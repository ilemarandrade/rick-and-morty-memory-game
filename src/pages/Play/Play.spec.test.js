import { useCharactersState } from "../../contexts/CharactersContext";
import { KEYSHOWINTRUCTIONS, setKey } from "../../utils/localStorage";
import { render, screen, waitFor } from "../../__test__/utils/customRender";
import Play from "./index";

describe("Play", () => {
  beforeEach(() => {
    jest.resetModules();
  });
  it("should render without errors and show the instructions", async () => {
    const { container } = render(<Play />);
    // snapshot testing
    await waitFor(() => {
      screen.getByText("Empezar");
    });
    expect(container).toMatchSnapshot();
  });
  it("should render without errors and game board", async () => {
    setKey({ key: KEYSHOWINTRUCTIONS, value: "not" });

    const WrapperToWaitCharacters = ({ children }) => {
      const { characters } = useCharactersState();
      return <>{characters && children}</>;
    };

    const { container } = render(
      <WrapperToWaitCharacters>
        <Play />
      </WrapperToWaitCharacters>
    );

    // snapshot testing
    await waitFor(() => {
      screen.getAllByText("Rick Sanchez");
    });

    expect(container).toMatchSnapshot();
  });
});
