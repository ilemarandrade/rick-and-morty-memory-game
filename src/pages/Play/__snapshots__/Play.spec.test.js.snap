// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Play should render without errors and game board 1`] = `
<div>
  <div
    class="d-flex jc-space-between"
  >
    <h2
      class="mb-1-5"
    >
      Aciertos: 0
    </h2>
    <h2
      class="mb-1-5"
    >
      Turnos: 0
    </h2>
  </div>
  <div
    class="root"
  >
    <button
      class="
      card-open card "
    >
      <img
        alt="Rick Sanchez"
        src="https://placekitten.com/200/200"
      />
      <p
        class="name"
        title="Rick Sanchez"
      >
        Rick Sanchez
      </p>
      <p
        class="origin"
      >
        Alive - Human
      </p>
    </button>
    <button
      class="
      card-open card "
    >
      <img
        alt="Rick Sanchez"
        src="https://placekitten.com/200/200"
      />
      <p
        class="name"
        title="Rick Sanchez"
      >
        Rick Sanchez
      </p>
      <p
        class="origin"
      >
        Alive - Human
      </p>
    </button>
    <button
      class="
      card-open card "
    >
      <img
        alt="Rick Sanchez"
        src="https://placekitten.com/200/200"
      />
      <p
        class="name"
        title="Rick Sanchez"
      >
        Rick Sanchez
      </p>
      <p
        class="origin"
      >
        Alive - Human
      </p>
    </button>
    <button
      class="
      card-open card "
    >
      <img
        alt="Rick Sanchez"
        src="https://placekitten.com/200/200"
      />
      <p
        class="name"
        title="Rick Sanchez"
      >
        Rick Sanchez
      </p>
      <p
        class="origin"
      >
        Alive - Human
      </p>
    </button>
    <button
      class="
      card-open card "
    >
      <img
        alt="Rick Sanchez"
        src="https://placekitten.com/200/200"
      />
      <p
        class="name"
        title="Rick Sanchez"
      >
        Rick Sanchez
      </p>
      <p
        class="origin"
      >
        Alive - Human
      </p>
    </button>
    <button
      class="
      card-open card "
    >
      <img
        alt="Rick Sanchez"
        src="https://placekitten.com/200/200"
      />
      <p
        class="name"
        title="Rick Sanchez"
      >
        Rick Sanchez
      </p>
      <p
        class="origin"
      >
        Alive - Human
      </p>
    </button>
    <button
      class="
      card-open card "
    >
      <img
        alt="Rick Sanchez"
        src="https://placekitten.com/200/200"
      />
      <p
        class="name"
        title="Rick Sanchez"
      >
        Rick Sanchez
      </p>
      <p
        class="origin"
      >
        Alive - Human
      </p>
    </button>
    <button
      class="
      card-open card "
    >
      <img
        alt="Rick Sanchez"
        src="https://placekitten.com/200/200"
      />
      <p
        class="name"
        title="Rick Sanchez"
      >
        Rick Sanchez
      </p>
      <p
        class="origin"
      >
        Alive - Human
      </p>
    </button>
    <button
      class="
      card-open card "
    >
      <img
        alt="Rick Sanchez"
        src="https://placekitten.com/200/200"
      />
      <p
        class="name"
        title="Rick Sanchez"
      >
        Rick Sanchez
      </p>
      <p
        class="origin"
      >
        Alive - Human
      </p>
    </button>
    <button
      class="
      card-open card "
    >
      <img
        alt="Rick Sanchez"
        src="https://placekitten.com/200/200"
      />
      <p
        class="name"
        title="Rick Sanchez"
      >
        Rick Sanchez
      </p>
      <p
        class="origin"
      >
        Alive - Human
      </p>
    </button>
    <button
      class="
      card-open card "
    >
      <img
        alt="Rick Sanchez"
        src="https://placekitten.com/200/200"
      />
      <p
        class="name"
        title="Rick Sanchez"
      >
        Rick Sanchez
      </p>
      <p
        class="origin"
      >
        Alive - Human
      </p>
    </button>
    <button
      class="
      card-open card "
    >
      <img
        alt="Rick Sanchez"
        src="https://placekitten.com/200/200"
      />
      <p
        class="name"
        title="Rick Sanchez"
      >
        Rick Sanchez
      </p>
      <p
        class="origin"
      >
        Alive - Human
      </p>
    </button>
  </div>
</div>
`;

exports[`Play should render without errors and show the instructions 1`] = `
<div>
  <h1
    class="ta-center"
  >
    Intrucciones
  </h1>
  <ul>
    <li>
      Tendrás 3 segundos para memorizar las cartas.
    </li>
    <li>
      Una vez volteadas puedes seleccionar 2 cartas.
    </li>
    <li>
      Si las cartas abiertas son iguales te sumará puntos a favor.
    </li>
    <li>
      Si las cartas son diferentes se colocarán como estaban y podrás volver a elegir otras dos más.
    </li>
    <li>
      Procura hacerlo la menor cantidad de intentos.
    </li>
    <li>
      Mientras menos intentos/turnos, más memoria tienes!
    </li>
  </ul>
  <div
    class="perfect-centered mb-1-5"
  >
    <input
      id="miCheckbox"
      type="checkbox"
    />
    <label
      for="miCheckbox"
    >
      No volver a ver las instrucciones
    </label>
  </div>
  <div
    class="perfect-centered"
  >
    <button
      class="button-base button-color-primary"
      color="primary"
    >
      Empezar
    </button>
  </div>
</div>
`;
